File: java/raster-function-service/src/main/java/com/esri/arcgisruntime/sample/rasterfunctionservice/MainActivity.java
/* Copyright 2018 Esri
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.esri.arcgisruntime.sample.rasterfunctionservice;

import java.util.List;

import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.widget.Button;

import com.esri.arcgisruntime.geometry.Point;
import com.esri.arcgisruntime.layers.RasterLayer;
import com.esri.arcgisruntime.loadable.LoadStatus;
import com.esri.arcgisruntime.mapping.ArcGISMap;
import com.esri.arcgisruntime.mapping.Basemap;
import com.esri.arcgisruntime.mapping.view.MapView;
import com.esri.arcgisruntime.raster.ImageServiceRaster;
import com.esri.arcgisruntime.raster.Raster;
import com.esri.arcgisruntime.raster.RasterFunction;
import com.esri.arcgisruntime.raster.RasterFunctionArguments;

public class MainActivity extends AppCompatActivity {

  private MapView mMapView;
  private Button mRasterFunctionButton;
  private ArcGISMap map;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);

    // inflate MapView from layout
    mMapView = findViewById(R.id.mapView);
    mRasterFunctionButton = findViewById(R.id.rasterButton);
    mRasterFunctionButton.setEnabled(false);
    // create a map with the BasemapType topographic
    map = new ArcGISMap(Basemap.createDarkGrayCanvasVector());
    final ImageServiceRaster imageServiceRaster = new ImageServiceRaster(getString(R.string.image_service_raster_url));
    final RasterLayer imageRasterLayer = new RasterLayer(imageServiceRaster);
    map.getOperationalLayers().add(imageRasterLayer);
    imageRasterLayer.addDoneLoadingListener(() -> {
      if (imageRasterLayer.getLoadStatus() == LoadStatus.LOADED) {
        // get the center point
        Point centerPnt = imageServiceRaster.getServiceInfo().getFullExtent().getCenter();
        mMapView.setViewpointCenterAsync(centerPnt, 55000000);
        mRasterFunctionButton.setEnabled(true);
      }
    });

    mRasterFunctionButton.setOnClickListener(v -> applyRasterFunction(imageServiceRaster));

    // set the map to be displayed in this view
    mMapView.setMap(map);
  }

  private void applyRasterFunction(Raster raster) {
    // create raster function from json string
    RasterFunction rasterFunction = RasterFunction.fromJson(getString(R.string.hillshade_simplified));
    // get parameter name value pairs used by hillside
    RasterFunctionArguments rasterFunctionArguments = rasterFunction.getArguments();
    // get a list of raster names associated with a raster function
    List<String> rasterName = rasterFunctionArguments.getRasterNames();
    rasterFunctionArguments.setRaster(rasterName.get(0), raster);
    // create raster as raster layer
    raster = new Raster(rasterFunction);
    RasterLayer hillshadeLayer = new RasterLayer(raster);
    map.getOperationalLayers().add(hillshadeLayer);
  }

  @Override
  protected void onPause() {
    super.onPause();
    mMapView.pause();
  }

  @Override
  protected void onResume() {
    super.onResume();
    mMapView.resume();
  }

  @Override
  protected void onDestroy() {
    super.onDestroy();
    mMapView.dispose();
  }
}

File: kotlin/raster-function-service/src/main/java/com/esri/arcgisruntime/samples/rasterfunctionservice/MainActivity.kt
/* Copyright 2017 Esri
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.esri.arcgisruntime.samples.rasterfunctionservice

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import com.esri.arcgisruntime.layers.RasterLayer
import com.esri.arcgisruntime.loadable.LoadStatus
import com.esri.arcgisruntime.mapping.ArcGISMap
import com.esri.arcgisruntime.mapping.Basemap
import com.esri.arcgisruntime.raster.ImageServiceRaster
import com.esri.arcgisruntime.raster.Raster
import com.esri.arcgisruntime.raster.RasterFunction
import kotlinx.android.synthetic.main.activity_main.*


class MainActivity : AppCompatActivity() {

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_main)

    // create a Dark Gray Vectory BaseMap
    val map = ArcGISMap(Basemap.createDarkGrayCanvasVector())
    // set the map to be displayed in this view
    mapView.map = map
    // create image service raster as raster layer
    val imageServiceRaster = ImageServiceRaster(
      resources.getString(R.string.image_service_raster_url)
    )
    val imageRasterLayer = RasterLayer(imageServiceRaster)
    map.operationalLayers.add(imageRasterLayer)
    // zoom to the extent of the raster service
    imageRasterLayer.addDoneLoadingListener {
      if (imageRasterLayer.loadStatus == LoadStatus.LOADED) {
        // zoom to extent of raster
        val centerPnt = imageServiceRaster.serviceInfo.fullExtent.center
        mapView.setViewpointCenterAsync(centerPnt, 55000000.0)
        // update raster with simplified hillshade
        applyRasterFunction(imageServiceRaster)
      }
    }
  }

  /**
   * Apply a raster function on the given Raster
   *
   * @param raster Input raster to apply function
   */
  private fun applyRasterFunction(raster: Raster) {
    // create raster function from json string
    val rasterFunction = RasterFunction.fromJson(resources.getString(R.string.hillshade_simplified))
    // get parameter name value pairs used by hillshade
    val rasterFunctionArguments = rasterFunction.arguments
    // get list of raster names associated with raster function
    val rasterName = rasterFunctionArguments.rasterNames
    // set raster to the raster name
    rasterFunctionArguments.setRaster(rasterName[0], raster)
    // create raster as raster layer
    val raster = Raster(rasterFunction)
    val hillshadeLayer = RasterLayer(raster)
    // add hillshade raster
    mapView.map.operationalLayers.add(hillshadeLayer)
  }

  override fun onPause() {
    super.onPause()
    mapView.pause()
  }

  override fun onResume() {
    super.onResume()
    mapView.resume()
  }

  override fun onDestroy() {
    super.onDestroy()
    mapView.dispose()
  }
}
